source/base/Promise.ooc

```diff
@@ -0,0 +1,93 @@
+import threading/Thread
+import os/Time
+
+PromiseState: enum {
+	Running
+	Completed
+	Cancelled
+	Error
+}
+
+Promise: abstract class<T> {
+	_result: T
+	_state: PromiseState
+	state ::= this _state
+
+	init: func
+
+	wait: abstract func -> T { _result }
+
+	wait: virtual func ~timeout (seconds: Double) -> T { _result }
+
+	cancel: virtual func -> Bool { false }
+
+	_execute: virtual func
+}
+
+ThreadedPromise : class<T> extends Promise<T> {
```

Comment:
Elsewhere, we've written this as `class <T>`, so do that here as well. Magic should probably catch this. @thomasfanell ?